### PR TITLE
Improve Layer Count

### PR DESCRIPTION
**Pull request by aendeavor**
# Description

Lowers the layer count (by 18 layers) just by using `\`. Does not change anything else. Plain and simple.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change that does improve existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: Dockerfile
Comment (wernerfred) [inline Dockerfile:35]: Not needed anymore?

**Review by wernerfred — COMMENTED** [review]
Oh wow, didn't know that is possible tbh 🤣 

Did a quick research and found this page: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/

> In older versions of Docker, it was important that you minimized the number of layers in your images to ensure they were performant. The following features were added to reduce this limitation:
>
> Only the instructions RUN, COPY, ADD create layers. Other instructions create temporary intermediate images, and do not increase the size of the build

So do we really have a benefit or is this just a local number of layers that is decreased without any impact to the final image anyways?

**Comment by aendeavor** [comment]
> Oh wow, didn't know that is possible tbh rofl
> 
> Did a quick research and found this page: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/
> 
> > In older versions of Docker, it was important that you minimized the number of layers in your images to ensure they were performant. The following features were added to reduce this limitation:
> > Only the instructions RUN, COPY, ADD create layers. Other instructions create temporary intermediate images, and do not increase the size of the build
> 
> So do we really have a benefit or is this just a local number of layers that is decreased without any impact to the final image anyways?

I was wondering this as well when I just stumbled upon this with a fellow student. After what you've posted, I don't this _actually_ adds a performance benefit, but looking at <https://hub.docker.com/layers/mailserver/docker-mailserver/edge/images/sha256-1caa65c0f7f85428a145922181093304b20baac57ca8afbdf4d0fa1115e031aa?context=explore> for example, they do appear as a layer? 

We don't need to merge this at all, I just looked at the link above and thought: "Hey, maybe we can reduce layers here :)"

File: Dockerfile
Comment (aendeavor) [reply Dockerfile:35]: The `ENV`s can be ommited, the shell remains necessary and is now located at the very top of the file. We need `SHELL` as we're using some Bash features during the build, e.g. `[[ ... ]]`.

**Comment by wernerfred** [comment]
Then this might be more a "style" question. Imho i see no problem with all `ENV` and `LABEL` to appear as a single layer on dockerhub although there is no change in fact.

Is this possible for `ARG` as well?

**Comment by aendeavor** [comment]
> Then this might be more a "style" question. Imho i see no problem with all `ENV` and `LABEL` to appear as a single layer on dockerhub although there is no change in fact.
> 
> Is this possible for `ARG` as well?

You can apparently not do this for `ARG`. I had no problem with the style as is, but my fellow said he knows using `\` to concatenate the `LABEL` and `ENV` command to be good style.

File: Dockerfile
Comment (wernerfred) [reply Dockerfile:35]: Ah i missed that `SHELL` at the very top

**Review by wernerfred — APPROVED** [review]
LGTM 👍🏻 
From my perspective go ahead 🚀

**Comment by aendeavor** [comment]
I will merge #1989 first when CI passes, then I will merge this PR and then announce the feature freeze in the maintainers discussion 💯

**Comment by casperklein** [comment]
I think, this is just a docker hub thing. LABEL, ENV etc. do not add additional layers. They just end up in the manifest.

You can verify this for example with [dive](https://github.com/wagoodman/dive). Or take a look at how many layers are pulled with `docker pull`. There will be no difference after this merge.

I don't open the style discussion 😄 That is just personal preference, if you prefer `\` syntax. What I do for example from time to time is `grep ENV Dockerfile` to get all ENVs. That would not be possible anymore (in that easy form). No show-stopper. My preference would be to leave it as it is.

**Comment by aendeavor** [comment]
> I think, this is just a docker hub thing. LABEL, ENV etc. do not add additional layers. They just end up in the manifest.
> 
> You can verify this for example with [dive](https://github.com/wagoodman/dive). Or take a look at how many layers are pulled with `docker pull`. There will be no difference after this merge.
> 
> I don't open the style discussion smile That is just personal preference, if you prefer `\` syntax. What I do for example from time to time is `grep ENV Dockerfile` to get all ENVs. That would not be possible anymore (in that easy form). No show-stopper. My preference would be to leave it as it is.

I understand. As this was started as a real improvement when it comes to layer count, but now only is a style thing. I will close this:) I actually prefer the way it is ATM over this PR, but I figured in the beginning if there was a real benefit, it would be worth it.

**Comment by casperklein** [comment]
What I do, to ultimately "optimize" the layer count to 1:

```
FROM docker.io/debian:buster-slim as build
[..]
# build final image
FROM scratch
COPY --from=build / /
```

The only disadvantage of this approach is, that you lose the ability to use a base-image (debian). Now you have one image (1 layer) without dependency to an upstream image (debian).

**Pull request closed without merging**
